### DISCUSSION TITLE
Enabling tools to be executed several times per generation

### DISCUSSION THREAD

**Discussion by smonux** [open-webui/open-webui#5792] [General]
I have been writing a tool for code execution ( https://github.com/smonux/open-webui-docker-execution) which is was starting to look promising for my needs. In its current form runs code in a isolated docker container only linked to the host by a shared directory.

My plan was to implement a sort of poor man's data analyst, where I would deposit some csv's or sqlite's, or whatever, and the LLM would proactively clean, describe and interpret the data with little guidance, just a bit of elaborated system prompting. Hopefully, it wouldn't require expensive models to work but  the like of deepseek-coder, qwen-2.5, etc.., but even  if  I'd needed o1-preview to achieve that goal, it would be still useful.

The thing is I have realized that Tools **are only called once by generation**. I tried fooling the system by creating several identical functions `run_python_code1`, `run_python_code2`, etc... but that also doesn't work, because you can't even use two different tools at once. 

For example, if you have a tool that shows the time, and another that provides the location of the user, only once will be used even if both would be useful to answer the user's request.

![imagen](https://github.com/user-attachments/assets/37e3f47a-b172-447f-87d4-4c2df8fb915b)
(In the first generation I had both tools enabled. It only executed the date related).

What it would be nice is to be able to define in the Tool class a `max_runs_per_generation` (which would default to 1 if absent), so the tool could be called several times (at best not consecutive). Furthermore, IMO, unrelated tools should always be composable (if the llm chooses to use them).

Is this achievable with the current architecture? I could try to implement it, but I'd like to have an idea of how hard it is and if it would be included.

Thank you.

**Comment by smonux** [comment]
I have made a little proof of concept of the feature:

https://github.com/smonux/open-webui/commit/c250936684990893d49ff8c4985635b7c0c31f10

Now it can run two tools. In the first iteration has two avaliable and just one in the second.

![imagen](https://github.com/user-attachments/assets/ad787026-8c14-4013-a78f-25330fa3a5a3)

```
INFO  [open_webui.main] tools_function_calling_prompt='Available Tools: [{"name": "location", "description": "\\n        It returns the location of the user\\n        :return: the city or town where the user resides\\n        ", "parameters": {"type": "object", "properties": {}, "required": []}}, {"name": "get_current_date", "description": "\\n\\n        Get the current date.\\n\\n\\n        :return: The current date as a string.\\n        ", "parameters": {"type": "object", "properties": {}, "required": []}}]\nReturn an empty string if no tools match the query. If a function tool matches, construct and return a JSON object in the format {"name": "functionName", "parameters": {"requiredFunctionParamKey": "requiredFunctionParamValue"}} using the appropriate tool and its parameters. Only return the object and limit the response to the JSON object without additional text.'

INFO  [open_webui.main] tools_function_calling_prompt='Available Tools: [{"name": "location", "description": "\\n        It returns the location of the user\\n        :return: the city or town where the user resides\\n        ", "parameters": {"type": "object", "properties": {}, "required": []}}]\nReturn an empty string if no tools match the query. If a function tool matches, construct and return a JSON object in the format {"name": "functionName", "parameters": {"requiredFunctionParamKey": "requiredFunctionParamValue"}} using the appropriate tool and its parameters. Only return the object and limit the response to the JSON object without additional text.'

```

**Edit:** That change doesn't allow tools to be executed multiple times. It runs the tool a maximum of `max_runs` times, but with the same output, because the language model isn't aware of previous executions. To make it work as intended, some adjustments to the code would be necessary.

**Comment by tjbck** [comment]
Related! https://github.com/open-webui/open-webui/issues/4219

**Comment by smonux** [comment]
Closing this, since it's related to #4219
